File: 2016/Day5/main.go
package main

import (
	"crypto/md5"
	"encoding/hex"
	"strconv"
)

func findPassword(doorID string) string {
	password := ""
	index := 0

	for len(password) < 8 {
		toHash := doorID + strconv.Itoa(index)
		hash := md5.Sum([]byte(toHash))
		hashResult := hex.EncodeToString(hash[:])

		if hashResult[:5] == "00000" {
			password += string(hashResult[5])
		}

		index++
	}

	return password
}

func findPassword2(doorID string) string {
	password := make([]byte, 8)
	for i := range password {
		password[i] = '_'
	}
	index := 0
	filledPositions := 0

	for filledPositions < 8 {
		toHash := doorID + strconv.Itoa(index)
		hash := md5.Sum([]byte(toHash))
		hashResult := hex.EncodeToString(hash[:])

		if hashResult[:5] == "00000" {
			position := hashResult[5]
			if position >= '0' && position <= '7' {
				pos := int(position - '0')
				if password[pos] == '_' {
					password[pos] = hashResult[6]
					filledPositions++
				}
			}
		}

		index++
	}

	return string(password)
}

func main() {
	doorID := "ojvtpuvg"

	println("Part 1:", findPassword(doorID))
	println("Part 2:", findPassword2(doorID))
}

File: 2016/Day5/main.py
import hashlib

def find_password(door_id):
    password = ""
    index = 0

    while len(password) < 8:
        to_hash = door_id + str(index)

        hash_result = hashlib.md5(to_hash.encode()).hexdigest()

        if hash_result.startswith("00000"):
            password += hash_result[5]

        index += 1

    return password

def find_password2(door_id):
    password = ['_'] * 8
    index = 0
    filled_positions = 0

    while filled_positions < 8:
        to_hash = door_id + str(index)

        hash_result = hashlib.md5(to_hash.encode()).hexdigest()

        if hash_result.startswith("00000"):
            position = hash_result[5]
            if position.isdigit():
                position = int(position)
                if position < 8 and password[position] == '_':
                    password[position] = hash_result[6]
                    filled_positions += 1

        index += 1

    return "".join(password)

door_id = "ojvtpuvg"

print(f"Part 1: {find_password(door_id)}")
print(f"Part 2: {find_password2(door_id)}")
